src/sinks/util/unix.rs

```diff
@@ -0,0 +1,291 @@
+use crate::sinks::util::SinkExt;
+use crate::{
+    event::{self, Event},
+    sinks::{Healthcheck, RouterSink},
+    topology::config::SinkContext,
+};
+use bytes::Bytes;
+use futures::{
+    future, stream::iter_ok, try_ready, Async, AsyncSink, Future, Poll, Sink, StartSend,
+};
+use serde::{Deserialize, Serialize};
+use snafu::Snafu;
+use std::io;
+use std::path::PathBuf;
+use std::time::{Duration, Instant};
+use tokio::codec::{BytesCodec, FramedWrite};
+use tokio::timer::Delay;
+use tokio_retry::strategy::ExponentialBackoff;
+use tokio_uds::UnixStream;
+use tracing::field;
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(deny_unknown_fields)]
+pub struct UnixSinkConfig {
+    pub path: PathBuf,
+    pub encoding: Encoding,
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(rename_all = "snake_case")]
+pub enum Encoding {
+    Text,
+    Json,
+}
+
+impl UnixSinkConfig {
+    pub fn new(path: PathBuf) -> Self {
+        Self {
+            path,
+            encoding: Encoding::Text,
+        }
+    }
+
+    pub fn build(&self, cx: SinkContext) -> crate::Result<(RouterSink, Healthcheck)> {
+        let encoding = self.encoding.clone();
+
+        let sink = Box::new(
+            UnixSink::new(self.path.clone())
+                .stream_ack(cx.acker())
+                .with_flat_map(move |event| iter_ok(Some(encode_event(event, &encoding)))),
+        );
+        let healthcheck = unix_healthcheck(self.path.clone());
+
+        Ok((sink, healthcheck))
+    }
+}
+
+#[derive(Debug, Snafu)]
+enum HealthcheckError {
+    #[snafu(display("Connect error: {}", source))]
+    ConnectError { source: std::io::Error },
+}
+
+fn unix_healthcheck(path: PathBuf) -> Healthcheck {
+    // Lazy to avoid immediately connecting
+    let check = future::lazy(move || {
+        UnixStream::connect(&path)
+            .map(|_| ())
+            .map_err(|source| HealthcheckError::ConnectError { source }.into())
+    });
+
+    Box::new(check)
+}
+
+fn encode_event(event: Event, encoding: &Encoding) -> Bytes {
+    let log = event.into_log();
+
+    let mut bytes = match encoding {
+        Encoding::Json => serde_json::to_vec(&log.unflatten())
+            .map(Bytes::from)
+            .expect("Unable to encode event as JSON."),
```

Comment:
If the JSON encoding fails, we definitely want an error, but not to panic Vector. This should either return an error or display it and return `None` like the other `encode_event` does.

Reply:
Good point.  Switching to the existing impl from tcp.rs fixes this